### PR TITLE
Add meson as an alternative to autotools with corresponding CI

### PR DESCRIPTION
**Pull request by jluebbe**
This adds support for meson as an alternative build system to autotools.

### CODE REVIEW

File: libcomposefs/meson.build [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  'lcfs-mount.c',
+  'lcfs-mount.h',
+  'xalloc-oversized.h',
+])
+
+libcomposefs = shared_library('composefs', 
+  source_files,
+  c_args : composefs_hash_cflags + hidden_visibility_cflags,
+  dependencies : libcrypto_dep,
+  version : meson.project_version(), # FIXME: or libversion?
```
Comment (cgwalters) [inline libcomposefs/meson.build:27]: How does this relate to https://github.com/containers/composefs/pull/300/files#diff-30d8f6be6320feeacf686be94f48c70869b52630e01ea625f0f15adc0d57c3e4R16 ?

File: tools/meson.build [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+executable('mount.composefs',
+    'mountcomposefs.c',
+    dependencies : [libcomposefs_dep],
+    install : true,
+    install_dir : get_option('sbindir'),
+)
+
+executable('composefs-info',
+    ['composefs-info.c', '../libcomposefs/hash.c'],
+    c_args : ['-DUSE_OBSTACK=0', '-DTESTING=0', '-DUSE_DIFF_HASH=0'],
```
Comment (cgwalters) [inline tools/meson.build:20]: We should be able to reference a variable for this to dedup with the top of https://github.com/containers/composefs/pull/300/files#diff-4f12b63e85afe8c8819073d64da906f24d8fe7ad05e0cc1db6fab193413e4410 right?

**Review by cgwalters — APPROVED** [review]
Nice! Only superficially skimming this (and without knowing meson really...my path went autotools -> go+rust basically) it looks great! I'd be in favor of killing the autotools build actually.

The thing I'd like to cross-check at least manually (even in CI perhaps) is that the autotools and meson builds result in "identical" outputs...I am not aware of any reason they shouldn't actually be bit-for-bit compatible.

File: libcomposefs/meson.build (hunk 1 above)
Comment (jluebbe) [reply libcomposefs/meson.build:27]: I saw that you have different versions for the project itself and the libcomposefs.so links. The pkgconfig file also includes a version field. Which version should go where? :)

Comment (jluebbe) [reply libcomposefs/meson.build]: I've changed it to match what autotools produces (project version 1.0.4 in composefs.pc, libversion 1.2.0 for the symlinks).

**Comment by jluebbe** [comment]
I've noticed that autotools builds both a shared and a static library, so I've updated this PR to do that as well.

**Comment by cgwalters** [comment]
Tried this out, it's so much faster than autotools! Nice!

Doing a `diff -uNr` between the two builds, I see:

```
--- /tmp/composefs-autotools/usr/lib64/pkgconfig/composefs.pc   2024-06-28 09:50:22.074577014 -0400
+++ /tmp/composefs-meson/usr/lib64/pkgconfig/composefs.pc       2024-06-28 09:49:34.856786643 -0400
@@ -1,13 +1,10 @@
 prefix=/usr
-exec_prefix=${prefix}
-libdir=/usr/lib64
 includedir=${prefix}/include
+libdir=${prefix}/lib64
 
-Name: Composefs
+Name: composefs
 Description: library for generating and using composefs images
 Version: 1.0.4
-Requires: 
-Requires.private:  libcrypto
+Requires.private: libcrypto
 Libs: -L${libdir} -lcomposefs
-Libs.private: -lcrypto
 Cflags: -I${includedir}
```

Which looks totally fine to me. Digging into the differences via `diffoscope` on the generated binaries and shared libraries, everything I see looks like it's an artifact of either:

- The source paths I used were different, flowing into all the debuginfo
- Some default linker flags are different

But that's all fine.

Thanks so much for doing this! I'm still in favor of dropping autotools; will file a tracker issue where we can debate that.

**Pull request merged**